Add a warning for RRef serialization

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34931 Support using self as the destination in rpc.remote for builtin operators
* #34921 Fix dist autograd context Example block format
* #34919 Fix example block format in Distributed Optimizer API doc
* #34914 Fix example format in Distributed Autograd doc
* #34890 Minor fixes for RPC API docs
* #34888 Update descriptions for transmitting CUDA tensors
* #34887 Removing experimental tag in for RPC and adding experimental tag for RPC+TorchScript
* #34885 Adding warnings for async Tensor serialization in remote and rpc_async
* **#34884 Add a warning for RRef serialization**

Differential Revision: [D20491278](https://our.internmc.facebook.com/intern/diff/D20491278)

mrshenli committed:
Add a warning for RRef serialization

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -122,6 +122,11 @@ PyObject* rpc_init(PyObject* /* unused */) {
+              RRefs can only be serialized and deserialized by the RPC module.
+              Serializing and deserializing RRefs without RPC will lead to
+              errors.
+
```
Should we add some more details here mentioning that if we use torch.save(), jit.pickle() or python pickle on RRefs this would cause unintended consequences?

pritamdamania87 replied:
Also, I'm assuming we will remove this warning once https://github.com/pytorch/pytorch/pull/34689 and https://github.com/pytorch/pytorch/pull/34755 lands?

mrshenli replied:
> Should we add some more details here mentioning that if we use torch.save(), jit.pickle() or python pickle on RRefs this would cause unintended consequences?

Yes, let me add. 

> Also, I'm assuming we will remove this warning once #34689 and #34755 lands?

It might be better keeping this warning even after we land those two? Because otherwise, applications would only learn that through trial-error?

mrshenli committed:
Update on "Add a warning for RRef serialization"


Differential Revision: [D20491278](https://our.internmc.facebook.com/intern/diff/D20491278)

[ghstack-poisoned]

mrshenli committed:
Update on "Add a warning for RRef serialization"


Differential Revision: [D20491278](https://our.internmc.facebook.com/intern/diff/D20491278)

[ghstack-poisoned]
